metric/metric.go

```diff
@@ -0,0 +1,176 @@
+// Package metric provides Prometheus metrics collection and monitoring.
+package metric
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/mem"
+	"github.com/shirou/gopsutil/net"
+)
+
+// Metrics contains the Prometheus metrics server and registered custom metrics.
+type Metrics struct {
+	httpServer           *http.Server
+	config               Config
+	webSocketConnections prometheus.Gauge
+	webRTCConnections    prometheus.Gauge
+	cpuUsage             prometheus.Gauge
+	memoryUsage          prometheus.Gauge
+	networkUsage         *prometheus.GaugeVec
+}
+
+// Config defines the configuration for the metrics server.
+type Config struct {
+	Port int    // Port for metrics server
+	Path string // Path for metrics endpoint
+}
+
+// Default values for metrics configuration.
+const (
+	DefaultMetricsPort = 9090
+	DefaultMetricsPath = "/metrics"
+)
+
+// New creates a new Metrics instance with the specified configuration.
+func New(config Config) *Metrics {
+	return &Metrics{
+		config: config,
+		webSocketConnections: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "websocket_connections_total",
+			Help: "Current number of WebSocket connections.",
+		}),
+		webRTCConnections: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "webrtc_connections_total",
+			Help: "Current number of WebRTC connections.",
+		}),
+		cpuUsage: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "cpu_usage_percentage",
+			Help: "CPU usage percentage.",
+		}),
+		memoryUsage: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "memory_usage_bytes",
+			Help: "Current memory usage in bytes.",
+		}),
+		networkUsage: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "network_usage_bytes",
+			Help: "Current network usage in bytes.",
+		}, []string{"direction"}), // Direction: "inbound" or "outbound"
+	}
+}
+
+// RegisterMetrics registers custom metrics with Prometheus.
+func (m *Metrics) RegisterMetrics() {
+	prometheus.MustRegister(m.webSocketConnections)
+	prometheus.MustRegister(m.webRTCConnections)
+	prometheus.MustRegister(m.cpuUsage)
+	prometheus.MustRegister(m.memoryUsage)
+	prometheus.MustRegister(m.networkUsage)
+}
+
+// Start initializes and starts the metrics HTTP server.
+func (m *Metrics) Start(stop <-chan struct{}) {
+	m.httpServer = &http.Server{
+		Addr:              fmt.Sprintf(":%d", m.config.Port),
+		Handler:           promhttp.Handler(),
+		ReadHeaderTimeout: 5 * time.Second,
+	}
+
```

Comment:
Consider subscribing to the broker's events and publishing metrics as demonstrated below:

[Reference Code](https://github.com/Gopring/PDN-WebRTC/blob/main/media/media.go#L45)

### Example: Subscribe to Events and Publish Metrics

```go
// Subscribe to broker events
pushEvent := m.broker.Subscribe(broker.ClientMessage, broker.PUSH)
// (Note: The `connectedEvent` subscription is available on the coordinator branch and will be published soon)
// connectedEvent := c.broker.Subscribe(broker.Media, broker.CONNECTED)

for {
    var err error
    select {
    case event := <-pushEvent.Receive():
        // Example: Publish a metric for an increase in push requests
        // Metric.Publish(IncreasePushRequest)

    case event := <-connectedEvent.Receive():
        // Example: Publish a metric for increased server connections
        // Metric.Publish(IncreaseServerConnection)
    }

    // Handle any errors during event processing
    if err != nil {
        log.Printf("Failed to handle event in Media: %v", err)
    }
}
```